### PR TITLE
perf: index-driven pagination for getMessages (#132)

### LINKED ISSUE #132: perf: getMessages loads all channel messages into memory via .collect()
## Problem

`getMessages` in `apps/convex/functions/messaging/messages.ts` (lines 191-234) calls `.collect()` which loads **every message in a channel** into memory, then filters, sorts, and paginates in JavaScript. This is O(total_messages) per query instead of O(page_size).

Every time a user opens a chat, the backend scans the channel's entire message history — even if only showing the most recent 20 messages. As channels grow (hundreds to thousands of messages), this becomes increasingly expensive in both memory and latency.

### Current implementation (lines 191-234)

```typescript
// Loads ALL messages for the channel into memory
let allMessages = await ctx.db
  .query("chatMessages")
  .withIndex("by_channel_createdAt", (q) => q.eq("channelId", args.channelId))
  .collect();  // ⚠️ O(n) where n = total messages in channel

// In-memory filter: remove deleted, blocked-user, and thread-reply messages
let topLevelMessages = allMessages
  .filter((m) => !m.isDeleted && (!m.senderId || !blockedUserIds.has(m.senderId)) && !m.parentMessageId);

// In-memory sort by lastActivityAt (thread bump ordering)
topLevelMessages.sort((a, b) => {
  const aTime = a.lastActivityAt ?? a.createdAt;
  const bTime = b.lastActivityAt ?? b.createdAt;
  return bTime - aTime;
});

// In-memory cursor-based pagination
if (args.cursor) {
  const cursorIndex = topLevelMessages.findIndex((m) => m._id === args.cursor);
  if (cursorIndex >= 0) {
    topLevelMessages = topLevelMessages.slice(cursorIndex + 1);
  }
}
```

### Root cause

Messages are sorted by `lastActivityAt` (a mutable field — it changes when a thread gets a reply, causing the parent message to "bump" to the top). The existing index `by_channel_createdAt` sorts by `createdAt` which is immutable. Since there's no index on `lastActivityAt`, the code can't use Convex's index-driven pagination and must load everything into memory to sort.

### 1. Add index to schema (`apps/convex/schema.ts`)

Add a new index to `chatMessages`:
```typescript
.index("by_channel_lastActivityAt", ["channelId", "lastActivityAt"])
```

### 2. Backfill migration (`apps/convex/migrations/backfillLastActivityAt.ts`)

Create a one-time internal mutation that sets `lastActivityAt = createdAt` on all existing top-level messages (no `parentMessageId`) that don't have it. Follow the pattern in `apps/convex/migrations/addChannelSlugs.ts`.

```typescript
export const backfillLastActivityAt = internalMutation({
  handler: async (ctx) => {
    const messages = await ctx.db.query("chatMessages").collect();
    let updatedCount = 0;
    for (const msg of messages) {
      if (!msg.parentMessageId && msg.lastActivityAt === undefined) {
        await ctx.db.patch(msg._id, { lastActivityAt: msg.createdAt });
        updatedCount++;
      }
    }
    return { updatedCount, total: messages.length };
  },
});
```

**Why this is needed:** Without backfilling, messages with `undefined` `lastActivityAt` sort before all numeric values in the Convex index, making them invisible on the first page of results. After backfilling, every top-level message has `lastActivityAt` populated and sorts correctly.

**If message volume is large** (>8000 messages), batch the migration using `.paginate()` with a continuation cursor to stay within Convex's per-mutation write limits.

### 3. Rewrite `getMessages` query (`apps/convex/functions/messaging/messages.ts`, lines 191-234)

Replace the entire `.collect()` + in-memory sort/filter/paginate block with:

```typescript
// --- Cursor resolution ---
let cursorTime: number | undefined;
if (args.cursor) {
  const cursorMsg = await ctx.db.get(args.cursor as Id<"chatMessages">);
  if (cursorMsg) {
    cursorTime = cursorMsg.lastActivityAt ?? cursorMsg.createdAt;
  }
}

// --- Index-driven fetch with server-side filtering ---
const fetchLimit = limit + blockedUserIds.size + 1;
const candidates = await ctx.db
  .query("chatMessages")
  .withIndex("by_channel_lastActivityAt", (q) => {
    const base = q.eq("channelId", args.channelId);
    return cursorTime !== undefined ? base.lt("lastActivityAt", cursorTime) : base;
  })
  .order("desc")
  .filter((q) =>
    q.and(
      q.eq(q.field("isDeleted"), false),
      q.eq(q.field("parentMessageId"), undefined)
    )
  )
  .take(fetchLimit);

// --- Post-filter blocked users (tiny set, negligible cost) ---
const topLevelMessages = candidates.filter(
  (m) => !m.senderId || !blockedUserIds.has(m.senderId)
);

const hasMore = topLevelMessages.length > limit;
const pageMessages = topLevelMessages.slice(0, limit);
const cursor = pageMessages.length > 0
  ? pageMessages[pageMessages.length - 1]._id
  : undefined;

const chronologicalMessages = [...pageMessages].reverse();
return { messages: chronologicalMessages, hasMore, cursor };
```

### Why this works

| Aspect | Before (`.collect()`) | After (index + `.filter()` + `.take()`) |
|--------|----------------------|----------------------------------------|
| Memory | O(total_messages) | O(page_size) |
| Server-side filtering | None — all in JS | `isDeleted` and `parentMessageId` filtered by Convex before `.take()` |
| Sorting | In-memory sort | Index-driven (Convex sorts by `lastActivityAt` via the index) |
| Blocked users | In-memory filter on all messages | Post-filter on page-sized result only |
| Cursor | Find-by-ID in full array | Single `ctx.db.get()` to resolve timestamp, then index range bound |

### 4. No frontend changes needed

- `useMessages.ts`, `useChatMessages.ts`, and `usePrefetchChannel.ts` all pass cursors opaquely
- Response shape `{ messages, hasMore, cursor }` is unchanged
- Cursor remains a message ID string

### 5. Tests (`apps/convex/__tests__/messaging/messages.test.ts`)

All 36 existing tests should pass unchanged. Add:
- Test that a message without `lastActivityAt` (simulating pre-migration data) still appears correctly
- Test that deleted messages in the index range don't reduce the returned page size (verifying `.filter()` runs before `.take()`)

## Deployment order

1. Deploy schema (index) + migration file + query rewrite together
2. Run backfill: `npx convex run migrations/backfillLastActivityAt:backfillLastActivityAt`
3. After backfill completes, all queries use the new index correctly

## Files to modify

- `apps/convex/schema.ts` — add `by_channel_lastActivityAt` index
- `apps/convex/functions/messaging/messages.ts` — rewrite `getMessages` (lines 191-234 only)
- `apps/convex/migrations/backfillLastActivityAt.ts` — new file (migration)
- `apps/convex/__tests__/messaging/messages.test.ts` — add edge case tests

## Known edge cases

1. **Cursor timestamp ties**: Two messages with identical `lastActivityAt` (rare — millisecond precision). Convex secondary-sorts by `_id`, but `.lt()` would skip same-timestamp messages. Worst case: one message skipped during deep pagination. Acceptable tradeoff.
2. **Blocked users reducing page size**: `fetchLimit = limit + blockedUserIds.size + 1` accounts for this. Typically blockedUserIds.size is 0-5.
3. **Thread reply deletion**: Deleting a reply doesn't un-bump the parent's `lastActivityAt`. This is existing behavior and acceptable (the thread "had activity").

## Verification

- Run messaging tests: `pnpm vitest apps/convex/__tests__/messaging/messages.test.ts`
- Run backfill on dev deployment
- Open a chat channel in the app and verify messages load correctly
- Verify thread bump still works (reply to old message, it should appear at top)
- Check Convex dashboard for reduced document reads per query

### PR DESCRIPTION
**Pull request by lilseyi**
## Summary

- **Replaces `.collect()` with index-driven pagination** in `getMessages` — previously loaded ALL messages in a channel into memory on every page load, now reads only ~150 docs (3x page size) regardless of channel size
- **Adds `by_channel_lastActivityAt` index** to `chatMessages` for efficient descending scan by thread activity
- **Includes backfill migration** to set `lastActivityAt = createdAt` on existing top-level messages

## Benchmark Results

```
================================================================================
  getMessages — Document Reads (Page size: 50)
================================================================================
  Channel Size   | OLD reads    | NEW reads    | Savings
  ----------------------------------------------------------------------------
  50 msgs        | 51           | 51           | 0%
  200 msgs       | 201          | 150          | 25%
  500 msgs       | 501          | 150          | 70%
  1000 msgs      | 1001         | 150          | 85%
  500 (page 2)   | 501          | 150          | 70%

  Projected (100 users, 300 msgs/channel, 10 opens/day):
  OLD: 9.0M reads/month → NEW: 4.5M reads/month (50% reduction)
================================================================================
```

## Deployment Steps

1. Deploy schema + code (index is created automatically)
2. Run backfill migration:
   ```bash
   npx convex run migrations/backfillLastActivityAt:backfillLastActivityAt
   ```

## Notes

- Cursor format changed from `messageId` to `"timestamp:messageId"` — opaque to frontend, no mobile changes needed
- Offline cache (`messageCache`) only stores initial page with `cursor=undefined`, so new format is fully backwards-compatible
- Reactions are loaded separately via `getReactionsForMessages` and are unaffected

## Test plan

- [x] All 37 existing message tests pass
- [x] All 14 blocking tests pass
- [x] All 966 project tests pass
- [x] Benchmark test included (`getMessages-benchmark.test.ts`)
- [ ] Deploy to staging and verify chat loads correctly
- [ ] Run backfill migration on staging
- [ ] Verify pagination (scroll up to load older messages)
- [ ] Verify thread bump ordering (reply to old thread, confirm it moves to top)

Closes #132

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes the core `getMessages` pagination/query strategy and cursor format, which could cause missing/duplicate messages or pagination regressions if edge cases (same timestamp, filtered rows) are wrong; also introduces a schema index and a backfill migration that must be run safely in production.
> 
> **Overview**
> **Reworks `getMessages` to avoid loading entire channels into memory.** The query now paginates by scanning a new `chatMessages` index (`by_channel_lastActivityAt`) in descending order with controlled over-fetch, filtering out deleted/blocked users and thread replies while only reading enough documents to fill a page.
> 
> **Introduces a new cursor scheme and data backfill.** The pagination cursor changes from a message ID to an opaque `"timestamp:ids"` token to handle tie-breaking at identical `lastActivityAt` values, and a new migration `backfillLastActivityAt` is added to set `lastActivityAt = createdAt` for existing top-level messages.
> 
> **Ensures new messages participate in the index and adds validation tooling.** Several message insertion paths (reach-out cards and scheduled bot posts) now set `lastActivityAt`, and a new `getMessages` benchmark test is added to compare estimated document reads and latency before vs after the optimization.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 373ea9d0bc95891b352009beb25e155075fca749. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Review by lilseyi — COMMENTED** [review]
## PR Review: `perf: index-driven pagination for getMessages (#132)`

### Overall: Solid optimization, a few issues to address

The core approach is sound — replacing `.collect()` with index-driven pagination is the right call and the benchmark numbers are compelling.

### Issues

**1. Cursor tie-breaking bug (medium)**
Lines 262-272: The first batch uses `.lte("lastActivityAt", scanCursorTime)` but this can return messages with the *same* timestamp that were already seen on the previous page. The cursor skip logic at line 282 only skips the exact cursor message by `_id`, but **other messages with the same timestamp that appeared before the cursor message on the previous batch could be duplicated**.

Consider: if 3 messages have `lastActivityAt = 1000` and the cursor points to the 2nd one, the next page's `.lte(1000)` will return all 3 again. You skip the cursor message, but the 1st message (already seen) gets re-included.

**2. `hasMore` is overly eager (low)**
Line 332: `hasMore: hasMore || !exhausted` — if the first batch returns fewer than `fetchBatch` items but still has enough accepted messages, `exhausted` will be `true` but `hasMore` might still be incorrect. This won't cause data loss but could cause an unnecessary empty-result fetch from the frontend.

**3. Migration `.collect()` on all messages (medium)**
The backfill migration at `migrations/backfillLastActivityAt.ts:15` calls `.collect()` on the entire `chatMessages` table. For a large table, this will hit Convex's function timeout/memory limits. Consider paginating the migration (e.g., process in batches of 1000 with a cursor, or use Convex's built-in migration helper).

**4. Pre-backfill messages silently dropped (low)**
Line 290: `if (m.lastActivityAt === undefined) continue` — between deploying the code and running the backfill, any pre-existing messages without `lastActivityAt` will be invisible. The deployment steps mention running the backfill after deploy, but there's a window where messages disappear. Consider falling back to `createdAt` instead of skipping.

### Suggestions

- For issue #1, you could encode more info in the cursor (e.g., include the index position or use a stricter `lt` with the `_id` as secondary sort)
- For issue #3, Convex recommends paginated internal mutations for large backfills — see their migration patterns
- For issue #4, change the `undefined` check to a fallback: `const activityAt = m.lastActivityAt ?? m.createdAt` and use that for filtering/sorting

### What looks good

- Cursor is fully opaque to all frontend callers — format change is safe
- `lastActivityAt` is properly set on all production message creation paths (`sendMessage`, `sendSystemMessage`, CLI)
- Thread bump logic correctly updates parent `lastActivityAt`
- Test coverage is decent with 37 message tests + 14 blocking tests passing
- Benchmark test is a nice addition for documenting the improvement

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: apps/convex/functions/messaging/messages.ts [hunk 1]
```diff
@@ -231,40 +231,135 @@ export const getMessages = query({
...
+      scanCursorTime = lastTime;
+
+      candidates = await ctx.db
+        .query("chatMessages")
+        .withIndex("by_channel_lastActivityAt", (q) =>
+          q.eq("channelId", args.channelId).lte("lastActivityAt", scanCursorTime!)
+        )
+        .order("desc")
+        .take(fetchBatch);
+    }
```
[the remark this answers is not in the document]
Comment (lilseyi) [reply apps/convex/functions/messaging/messages.ts]: Fixed in 3494fae. Replaced boundary-only tracking with a `processedIds` set that tracks ALL processed message IDs across batches. Added no-progress detection: if a batch yields zero new accepted messages, we break instead of looping forever.

File: apps/convex/functions/messaging/messages.ts [hunk 2]
```diff
@@ -231,40 +231,119 @@ export const getMessages = query({
...
+      if (accepted.length > limit || candidates.length < fetchBatch) {
+        break;
+      }
+
+      // If no new messages were accepted in this batch and we processed
+      // all candidates, we're stuck at a timestamp with only filtered-out
+      // messages. Break to avoid an infinite loop.
+      if (accepted.length === acceptedBefore) {
+        break;
+      }
```
[the remark this answers is not in the document]
Comment (lilseyi) [reply apps/convex/functions/messaging/messages.ts]: Fixed in 373ea9d. The break condition now checks `newCandidateCount === 0` — only breaks when ALL candidates in a batch were already in `processedIds` (genuinely stuck). If new candidates were found but merely filtered out (deleted, blocked, thread replies), the loop continues scanning for valid messages further down the index.
